### ISSUE TITLE
SNN accuacy is less as compared to ANN accuracy

### ISSUE THREAD

**Issue by sauravtii** [fangwei123456/spikingjelly#259]
I am trying out this code (https://github.com/fangwei123456/spikingjelly/blob/master/spikingjelly/activation_based/ann2snn/examples/cnn_mnist.py). The only difference is I have changed the CNN architecture to VGG-16 and I am using CIFAR-10 dataset. Initially I trained it for 50 epochs and the ANN accuray was 94 % and SNN accuracy using Maxnorm was 71 % and 76 % by using Robustnorm. Can anyone please tell me how I can I get a good accuracy for SNN as well ?

**Comment by DingJianhao** [comment]
hi there, you can increase the timesteps to some bigger number like 500. MAXNORM and ROBUSTNORM are very primary methods yet useful. A small timestep may cause decrease in accuracy. Good luck to you!

**Comment by sauravtii** [comment]
Hi, I tried with 500 timesteps but still the accuracy is not good. I have provided the output below.

```
Epoch: 99
100%|██████████| 782/782 [00:07<00:00, 107.48it/s]
Validating Accuracy: 0.972

100%|██████████| 200/200 [00:01<00:00, 127.64it/s]
ANN Validating Accuracy: 0.8420
---------------------------------------------
Converting using MaxNorm
100%|██████████| 782/782 [00:07<00:00, 107.32it/s]
Simulating...
100%|██████████| 200/200 [05:29<00:00,  1.65s/it]
SNN accuracy (simulation 500 time-steps): 0.6975
---------------------------------------------
Converting using RobustNorm
100%|██████████| 782/782 [01:50<00:00,  7.10it/s]
Simulating...
100%|██████████| 200/200 [05:27<00:00,  1.64s/it]
SNN accuracy (simulation 500 time-steps): 0.6452
---------------------------------------------
Converting using 1/2 max(activation) as scales...
100%|██████████| 782/782 [00:07<00:00, 106.99it/s]
Simulating...
100%|██████████| 200/200 [05:28<00:00,  1.64s/it]
SNN accuracy (simulation 500 time-steps): 0.7526
---------------------------------------------
Converting using 1/3 max(activation) as scales
100%|██████████| 782/782 [00:07<00:00, 106.80it/s]
Simulating...
100%|██████████| 200/200 [05:28<00:00,  1.64s/it]
SNN accuracy (simulation 500 time-steps): 0.7735
---------------------------------------------
Converting using 1/4 max(activation) as scales
100%|██████████| 782/782 [00:07<00:00, 107.23it/s]
Simulating...
100%|██████████| 200/200 [05:28<00:00,  1.64s/it]
SNN accuracy (simulation 500 time-steps): 0.7336
---------------------------------------------
Converting using 1/5 max(activation) as scales
100%|██████████| 782/782 [00:07<00:00, 107.44it/s]
Simulating...
100%|██████████| 200/200 [05:51<00:00,  1.76s/it]
SNN accuracy (simulation 500 time-steps): 0.6390

```

**Comment by DingJianhao** [comment]
well, VGG-16, which pooling method do you choose?

**Comment by sauravtii** [comment]
Hi, I am using max pooling

**Comment by fangwei123456** [comment]
Try to use the average pooling.

**Comment by sauravtii** [comment]
> could you please share your code, I'm having the same problem as you

My VGG-16 architecture:
```
class CNN(nn.Module):
    def __init__(self):
        super().__init__()
        self.network = nn.Sequential(
            nn.Conv2d(3, 64, 3, 1, 1),     # conv-1
            nn.BatchNorm2d(64),
            nn.ReLU(),

            nn.Conv2d(64, 64, 3, 1, 1),    # conv-2
            nn.BatchNorm2d(64),
            nn.ReLU(),
            # nn.MaxPool2d(2, 2),                        
            nn.AvgPool2d(2, 2),

            nn.Conv2d(64, 128, 3, 1, 1),   # conv-3
            nn.BatchNorm2d(128),
            nn.ReLU(),

            nn.Conv2d(128, 128, 3, 1, 1),   # conv-4
            nn.BatchNorm2d(128),
            nn.ReLU(),
            # nn.MaxPool2d(2, 2),                        
            nn.AvgPool2d(2, 2),

            nn.Conv2d(128, 256, 3, 1, 1),   # conv-5
            nn.BatchNorm2d(256),
            nn.ReLU(),

            nn.Conv2d(256, 256, 3, 1, 1),   # conv-6
            nn.BatchNorm2d(256),
            nn.ReLU(),
            
            nn.Conv2d(256, 256, 3, 1, 1),   # conv-7
            nn.BatchNorm2d(256),
            nn.ReLU(),
            # nn.MaxPool2d(2, 2),                        
            nn.AvgPool2d(2, 2),

            nn.Conv2d(256, 512, 3, 1, 1),   # conv-8
            nn.BatchNorm2d(512),
            nn.ReLU(),

            nn.Conv2d(512, 512, 3, 1, 1),   # conv-9
            nn.BatchNorm2d(512),
            nn.ReLU(),

            nn.Conv2d(512, 512, 3, 1, 1),   # conv-10
            nn.BatchNorm2d(512),
            nn.ReLU(),
            # nn.MaxPool2d(2, 2),                        
            nn.AvgPool2d(2, 2),

            nn.Conv2d(512, 512, 3, 1, 1),   # conv-11
            nn.BatchNorm2d(512),
            nn.ReLU(),

            nn.Conv2d(512, 512, 3, 1, 1),   # conv-12
            nn.BatchNorm2d(512),
            nn.ReLU(),

            nn.Conv2d(512, 512, 3, 1, 1),   # conv-13
            nn.BatchNorm2d(512),
            nn.ReLU(),
            # nn.MaxPool2d(2, 2),            
            nn.AvgPool2d(2, 2),

            nn.Flatten(),
            
            nn.Linear(1*1*512, 4096),

            nn.Linear(4096, 4096),

            nn.Linear(4096, 10)
            
        )

    def forward(self,x):
        x = self.network(x)
        return x

```

**Comment by DingJianhao** [comment]
great，change to avgpool and your problem may be solved! @sauravtii

**Comment by sauravtii** [comment]
Thanks @DingJianhao. The accuracy did improve. But is there any way that I can get more than 90 % accuracy ?
I have provided my output:

```
Epoch: 149
100%|██████████| 391/391 [00:06<00:00, 60.81it/s]
Validating Accuracy: 0.979

100%|██████████| 200/200 [00:01<00:00, 134.99it/s]
ANN Validating Accuracy: 0.8489
---------------------------------------------
Converting using MaxNorm
100%|██████████| 391/391 [00:06<00:00, 64.59it/s]
Simulating...
100%|██████████| 200/200 [05:22<00:00,  1.61s/it]
SNN accuracy (simulation 500 time-steps): 0.8379
---------------------------------------------
Converting using RobustNorm
100%|██████████| 391/391 [02:05<00:00,  3.12it/s]
Simulating...
100%|██████████| 200/200 [05:23<00:00,  1.62s/it]
SNN accuracy (simulation 500 time-steps): 0.8060
---------------------------------------------
Converting using 1/2 max(activation) as scales...
100%|██████████| 391/391 [00:06<00:00, 64.59it/s]
Simulating...
100%|██████████| 200/200 [05:22<00:00,  1.61s/it]
SNN accuracy (simulation 500 time-steps): 0.8433
---------------------------------------------
Converting using 1/3 max(activation) as scales
100%|██████████| 391/391 [00:06<00:00, 65.04it/s]
Simulating...
100%|██████████| 200/200 [05:24<00:00,  1.62s/it]
SNN accuracy (simulation 500 time-steps): 0.8300
---------------------------------------------
Converting using 1/4 max(activation) as scales
100%|██████████| 391/391 [00:05<00:00, 65.21it/s]
Simulating...
100%|██████████| 200/200 [05:23<00:00,  1.62s/it]
SNN accuracy (simulation 500 time-steps): 0.7332
---------------------------------------------
Converting using 1/5 max(activation) as scales
100%|██████████| 391/391 [00:06<00:00, 64.90it/s]
Simulating...
100%|██████████| 200/200 [05:25<00:00,  1.63s/it]
SNN accuracy (simulation 500 time-steps): 0.4973
```

**Comment by DingJianhao** [comment]
try new algos
I recommend you read [Optimized Potential Initialization for Low-latency Spiking Neural Networks](https://arxiv.org/abs/2202.01440)

**Comment by sauravtii** [comment]
Sure, I will read that. Thank you for your help!

**Comment by AllenYolk** [comment]
ANN → SNN 转换后的精度损失通常来自几个常见原因（按影响从大到小）：

1. **激活值分布不匹配**。原 ReLU 网络的激活值分布与转换后脉冲发放率分布差距较大。可以用 RobustNorm：

   ```python
   from spikingjelly.activation_based.ann2snn import modules
   snn_model = modules.Converter(mode="robustnorm")
   ```

2. **最大池化 vs 平均池化**。AvgPool 转换后信息损失更小，能保留更多细节，**强烈推荐**把 `MaxPool2d` 换成 `AvgPool2d` 再做转换。

3. **时间步 T 不够**。T 较小时 spike rate 的离散化误差明显。先用 T=500 评估上界，再视情况往下调。

4. **bias 项**。转换算法只处理 `Conv/Linear + BN`，若有 `nn.ReLU` 之外的非线性（bias 偏移等），可能引入误差。

调参顺序建议：先 AvgPool → 再 RobustNorm / MaxNorm → 再加大 T。如果 train acc 接近 ANN 但 test acc 差距大，更可能是过拟合而不是转换问题。

参考：
- 教程：https://spikingjelly.readthedocs.io/en/latest/activation_based/5_ann2snn.html
- 推荐阅读：Optimized Potential Initialization for Low-latency SNNs (https://arxiv.org/abs/2205.03286)

**Comment by AllenYolk** [comment]
The most common causes of ANN→SNN accuracy loss (in order of impact):

1. **Activation distribution mismatch**. ReLU activations of the ANN often don't map cleanly to SNN spike rates. Use `RobustNorm`:

   ```python
   from spikingjelly.activation_based.ann2snn import modules
   snn_model = modules.Converter(mode="robustnorm")
   ```

2. **Max-pool vs Avg-pool**. AvgPool loses less information under conversion — **strongly recommended** to swap `MaxPool2d` for `AvgPool2d` before converting.

3. **Insufficient `T`**. With small `T`, the discrete spike-rate quantization error dominates. Try `T=500` first to find the ceiling, then dial back.

4. **Bias terms**. The ANN2SNN algorithm only handles `Conv/Linear + BN`; any extra non-linearity or bias offset introduces error.

Suggested order: AvgPool → RobustNorm/MaxNorm → increase `T`. If train acc is close to ANN but test acc is much lower, that's more likely overfitting than a conversion problem.

References:
- Tutorial: https://spikingjelly.readthedocs.io/en/latest/activation_based/5_ann2snn.html
- Recommended reading: Optimized Potential Initialization for Low-latency SNNs (https://arxiv.org/abs/2205.03286)
